src/libraries/System.Private.CoreLib/src/System/Numerics/BFloat16.cs

```diff
@@ -0,0 +1,2011 @@
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(short value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="Half" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(Half value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="int" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(int value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="long" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(long value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="Int128" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(Int128 value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="nint" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(nint value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="float" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(float value)
+        {
+            uint bits = BitConverter.SingleToUInt32Bits(value);
+            uint upper = bits >> 16;
+            // Only do rounding for finite numbers
+            if (float.IsFinite(value))
```

Comment:
Is this branch beneficial? It is basically skipping `Infinity` (never any payload) and `NaN` (normally no payload) and so is introducing an additional branch for the common path (finite values).

In a heavy loop with mostly finites, an `Infinity` would then be mispredicted and incur roughly a 20 cycle penalty which is more than the cost of just doing the rounding to get back the same result

Reply:
Ah right, because the rounding logic here can overflow the exponent as well, thus turn `NaN` into `-0.0`.